src/main/java/org/opentripplanner/routing/graph/Graph.java

```diff
@@ -674,7 +674,14 @@ public FeedScopedId getOrCreateServiceIdForDate(ServiceDate serviceDate) {
         FeedScopedId serviceId = ((CalendarServiceImpl)getCalendarService()).getOrCreateServiceIdForDate(serviceDate);
 
         if (!serviceCodes.containsKey(serviceId)) {
-            serviceCodes.put(serviceId, serviceCodes.size());
+            // Calculating new unique serviceCode based on size (!)
+            final int serviceCode = serviceCodes.size();
+            serviceCodes.put(serviceId, serviceCode);
+
+            if (index.getServiceCodesRunningForDate().containsKey(serviceDate)) {
+                // Need to add to serviceCodeRunningForDate-index so it is included in trip-requests
+                index.getServiceCodesRunningForDate().get(serviceDate).add(serviceCode);
+            }
```

Comment:
Would it be preferable to create a new set if one doesn't already exist, to make sure the `serviceCode` is always added?

```suggestion
            index.getServiceCodesRunningForDate()
                    .computeIfAbsent(serviceDate, (ignored) -> new TIntHashSet())
                    .add(serviceCode);
```

Reply:
I think not, the reason is that OTP should route for a continuous set of dates, like 2021-06-25 to 2021-12-25. Adding dates outside this period might have side-effects (at least possible memory leakage for an open date interval). This make sure we do not add new dates as part of a RT update. If we open for it, then this fix become a lot more riski, and needs more investigation.

Reply:
So, what we need to do if we should add a new date is t check the configured date range, but I this this fix is an critical fix, and if someone need RT updates to add Patterns outside the existing time schedule, that is a new PR.

Reply:
I am concerned about the thread safety here with one thread writing and multiple threads reading the same map - but again this is a critical fix, and fixing the concurrency issue is another PR - it did exist before this fix as well.

Reply:
Just to clarify --  if my reading of the code is correct `serviceCodesRunningForDate` only contains the dates for which there is an active service in the scheduled data. And so the date may be within the continuous set of dates, but still not have an entry.

https://github.com/opentripplanner/OpenTripPlanner/blob/5a8af0cbb76ee24c37b12f439e79f98a865920ab/src/main/java/org/opentripplanner/routing/graph/GraphIndex.java#L136-L151

I don't want to hold up a critical fix, so as far as I'm concerned it may be merged as is.

Reply:
Yes you are right there might be holes in the range, but it is not very likely. And no rush, we are rolling this out independently of merging it upstream. So maybe we should check the date is in range, also we should not generate serviceID if the date is outside the range. I can patch it.